jayden/implement-base-node-with-enum-based-static-casting

BayernMuller opened the pull request.

BayernMuller committed:
feat: Add BaseNode class for DOM structure representation

- Introduced BaseNode class to serve as a foundational element in the DOM, supporting node types and parent-child relationships.
- Implemented methods for managing node properties, including text content and input/output connections, with assertions for parameter validation.
- Enhanced memory management by utilizing weak_ptr for parent references.

BayernMuller committed:
feat: Introduce TagNode and TextNode classes for DOM representation

- Added TagNode class to represent HTML tags, including methods for managing child nodes and attributes.
- Introduced TextNode class for representing text nodes within the DOM.
- Updated HtmlToken structure to include attributes, classes, and an ID for enhanced HTML token management.

BayernMuller committed:
feat: Enhance BaseNode class with static assertions for type safety

- Added static_assert to ensure that the template type T is a derived class of BaseNode.
- Included an assertion to validate node type consistency during type casting in the As() method.

BayernMuller committed:
fix: Add missing newlines at the end of header files for consistency

- Ensured all header files in the DOM and utils directories end with a newline for better compatibility and adherence to coding standards.

Jin-uu commented on src/dom/base_node.hpp:
```diff
@@ -0,0 +1,89 @@
+  template <typename T>
+  T* As() const noexcept {
+    static_assert(std::is_base_of<BaseNode, T>::value, "T must be a derived class of BaseNode");
+    ARBORIS_ASSERT(T::kNodeType == node_type_, "Node type mismatch. got " << static_cast<std::uint32_t>(node_type_)
+                                                                          << " expected "
+                                                                          << static_cast<std::uint32_t>(T::kNodeType));
+    return (T::kNodeType == node_type_) ? static_cast<T*>(this) : nullptr;
+  }
```
```
ARBORIS_ASSERT(T::kNodeType == node_type_, "Node type mismatch. got " << static_cast<std::uint32_t>(node_type_)
                                                                      << " expected "
                                                                      << static_cast<std::uint32_t>(T::kNodeType));
```
요 부분 존재 의의를 잘 모르겠음.
`As`함수 목적이 `TextNode` <-> `TagNode` 상호간 static cast일텐데, 이 경우 Debug에서 assert할 이유가 뭐지?

BayernMuller replied:
`As` 는 `Base* ptr = new Derived` 에서 `Base*` 를 `Derived *` 로 바꾸는 함수입니다. 여기서 `Base`의 다른 자식 클래스가 `Derived*` 가 되는것을 막는거에요

Jin-uu commented on src/dom/tag_node.hpp:
```diff
@@ -0,0 +1,48 @@
+
+  const std::vector<std::string>& classes() const noexcept { return html_token_.classes; }
+
+  std::string_view id() const noexcept { return html_token_.id; }
```
`[[nodiscard]]` 추가하는게 좋을 듯?

BayernMuller replied:
이게 clang-tidy 에서 지시를 안 하더라고여. 기준이 뭔지 알아보겠습니다

Jin-uu commented on src/dom/tag_node.hpp:
```diff
@@ -0,0 +1,48 @@
+
+  std::string_view id() const noexcept { return html_token_.id; }
+
+  void AddChild(std::shared_ptr<BaseNode> child) {
+    ARBORIS_ASSERT(child != nullptr, "child must not be nullptr.");
+    children_.emplace_back(std::move(child));
+  }
```
Debug 모드 아닐 때 검사 안해도 되려나?

BayernMuller replied:
정상적인 경우에는 DomBuilder 에서 절대로 nullptr 로 호출 안 시킬거라, 릴리즈 모드일때는 그럴일이 없을 것 같아요

릴리즈때 그럴 경우를 가정해서 널체킹 하는 것 보다, Debug 레벨에서 확실히 잡고 넘어가는게 좋을 것 같아요

Jin-uu commented on src/dom/text_node.hpp:
```diff
@@ -0,0 +1,30 @@
+
+ public:
+  explicit TextNode(std::uint32_t node_id, std::shared_ptr<BaseNode> parent = nullptr)
+      : BaseNode(kNodeType, node_id, std::move(parent)) {}
```
`TextNode`가 루트인 경우도 생각 해야되려나.
정상적인 HTML이면 `TagNode`가 루트가 되어야 할 듯 한데, `DomBuilder`에서 검사하는게 나으려나?

BayernMuller replied:
TextNode 는 어차피 children 이 없어서 괜찮을 듯여

Jin-uu reviewed (commented):
`static_cast` 방식이 좀 더 구조가 깔끔한 듯. 현 PR로 결정합시다.

Node 클래스 관련해서 test는 별도 PR로 뺼까?

BayernMuller committed:
refactor: Update TextNode constructor to require a parent parameter

- Modified the TextNode constructor to remove the default value for the parent parameter, ensuring that a parent node must be explicitly provided during instantiation.

BayernMuller committed:
refactor: Add [[nodiscard]] attribute to TagNode methods for improved compiler warnings

- Enhanced the TagNode class by adding the [[nodiscard]] attribute to methods returning references and views, promoting better usage practices and compiler warnings for unused return values.

BayernMuller committed:
refactor: Update DOM structure by replacing Node with BaseNode and introducing TagNode and TextNode

- Refactored the DOMBuilder to utilize BaseNode instead of Node, enhancing type safety and flexibility.
- Introduced TagNode and TextNode classes, ensuring proper parent-child relationships within the DOM.
- Updated CMakeLists.txt to include new header files for the refactored structure.

BayernMuller committed:
refactor: Enhance BaseNode and DOMBuilder with type safety and new methods

- Added a node_type() method to BaseNode for improved type identification.
- Introduced a const version of the As() method in BaseNode for better const-correctness.
- Updated DOMBuilder to utilize the new node_type() method for parent node checks.
- Added a tag() method to TagNode for accessing the associated HTML token.
- Removed redundant static constexpr declarations from TagNode and TextNode for cleaner code.

BayernMuller committed:
refactor: Simplify static assertions in BaseNode's As() method

- Updated static_assert usage in the As() method to utilize std::is_base_of_v for improved readability and conciseness, ensuring T is a derived class of BaseNode.

BayernMuller committed:
refactor: Improve DOMBuilder structure and logic

- Updated DOMBuilder to initialize root_ with a TagNode, enhancing the DOM structure.
- Simplified parent node handling in FeedOpenToken by directly adding children to the parent.
- Refined the close token validation logic to utilize the tag() method from TagNode for better clarity.

BayernMuller committed:
refactor: Remove redundant assertion in DOMBuilder's FeedCloseToken

- Eliminated the assertion checking if the top node is a tag node in FeedCloseToken, as the node type is already validated through the updated structure.

BayernMuller committed:
refactor: Update parent type in BaseNode, TagNode, and TextNode for improved type safety

- Changed the parent parameter type in BaseNode, TagNode, and TextNode constructors from std::shared_ptr<BaseNode> to std::shared_ptr<TagNode>, enhancing type safety and ensuring proper parent-child relationships within the DOM.
